src/Castle.Core/DynamicProxy/CustomAttributeInfo.cs

```diff
@@ -0,0 +1,294 @@
+			FieldInfo[] namedFields,
+			object[] fieldValues)
+			: this(con, constructorArgs, EmptyProperties, EmptyValues, namedFields, fieldValues)
+		{
+		}
+
+		public CustomAttributeInfo(
+			ConstructorInfo con,
+			object[] constructorArgs)
+			: this(con, constructorArgs, EmptyProperties, EmptyValues, EmptyFields, EmptyValues)
+		{
+		}
+
+		public static CustomAttributeInfo FromExpression(Expression<Func<Attribute>> expression)
+		{
+			var namedProperties = new List<PropertyInfo>();
+			var propertyValues = new List<object>();
+			var namedFields = new List<FieldInfo>();
+			var fieldValues = new List<object>();
+
+			var newExpression = expression.Body as NewExpression;
+			if (newExpression == null)
+			{
+				var memberInitExpression = expression.Body as MemberInitExpression;
+				if (memberInitExpression == null)
+				{
+					throw new ArgumentException("The expression must be either a simple constructor call or an object initializer expression");
+				}
+
+				newExpression = memberInitExpression.NewExpression;
+
+				foreach (var binding in memberInitExpression.Bindings)
+				{
+					var assignment = binding as MemberAssignment;
+					if (assignment == null)
+					{
+						throw new ArgumentException("Only assignments bindings are supported");
+					}
+
+					object value = GetAttributeArgumentValue(assignment.Expression, true);
+
+					var property = assignment.Member as PropertyInfo;
+					if (property != null)
+					{
+						namedProperties.Add(property);
+						propertyValues.Add(value);
+					}
+					else
+					{
+						var field = assignment.Member as FieldInfo;
+						if (field != null)
+						{
+							namedFields.Add(field);
+							fieldValues.Add(value);
+						}
+						else
+						{
+							throw new ArgumentException("Only property and field assignments are supported");
+						}
+					}
+				}
+			}
+
+			var ctorArguments = new List<object>();
+			foreach (var arg in newExpression.Arguments)
+			{
+				object value = GetAttributeArgumentValue(arg, true);
+				ctorArguments.Add(value);
+			}
+
+			return new CustomAttributeInfo(
+				newExpression.Constructor,
+				ctorArguments.ToArray(),
+				namedProperties.ToArray(),
+				propertyValues.ToArray(),
+				namedFields.ToArray(),
+				fieldValues.ToArray());
+		}
+
+		private static object GetAttributeArgumentValue(Expression arg, bool allowArray)
+		{
+			var constant = arg as ConstantExpression;
+			if (constant == null)
+			{
+				if (allowArray)
+				{
+					var newArrayExpr = arg as NewArrayExpression;
+					if (newArrayExpr != null)
+					{
+						var array = new object[newArrayExpr.Expressions.Count];
+						int index = 0;
+						foreach (var expr in newArrayExpr.Expressions)
+						{
+							object value = GetAttributeArgumentValue(expr, false);
+							array[index] = value;
+							index++;
+						}
+						return array;
+					}
+				}
+				throw new ArgumentException("Only constant and single-dimensional array expressions are supported");
```

Comment:
> it can only happen if we find an array inside an array

Ah! Yes. Thanks. I see it now after rereading. And having you tell me. :blush:

>  I can rename it and/or use a named argument on the call site to make things clearer if necessary.

The meaning of the parameter is quite clear on the inside of the method. And I like the name.
For me, a named argument would improve the call sites. I know they're not everyone's [cup of tea](http://idioms.thefreedictionary.com/not+cup+of+tea), though.

An alternative would be a microtype (probably an enum) so we could have explicitly named values at the call sites, but again, maybe that's too much.

In the end, it's a private method, and not hard for anyone who's confused by the calls to look up the meaning of the bool. Thanks for listening.